docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,340 @@
+            ub = dt5 * (ut[0, -1, 0] + ut)​
+
+    @gtscript.stencil(backend=backend_type)​
+    def y_edge_ub(ut: FIELD_TYPE, ub: FIELD_TYPE, *, dt4: float):
+        with computation(PARALLEL), interval(...):
+            ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
+
+    # Requires new python function to call stencils on each region
+    def ubke(uc, vc, ut, ub, dt5, dt4, grid):​
+        domain_y_edge = (grid.ni, 1, grid.nz)
+        domain_x_edge = (1, grid.nj, grid.nz)
+        main_ub(uc, vc, grid.cosa, grid.rsina, ub, dt5=dt5, ​
+                origin=(grid.local_istart, grid.local_jstart, 0),
+                domain=(grid.ni, grid.nj, grid.nz))​
+        if grid.west_edge:​
+            x_edge_ub(ut, ub, dt5=dt5, ​origin=(grid.local_istart, grid.local_jstart, 0), ​domain=domain_x_edge)​
+        if grid.south_edge:
+            y_edge(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jstart, 0), domain=domain_y_edge)
+        if grid.north_edge:
+            y_edge_ub(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jend, 0), domain=domain_y_edge)
+        if grid.east_edge:
+            x_edge_ub(ut, ub, dt5=dt5, origin=(grid.local_iend, grid.local_jstart, 0), domain=domain_x_edge)
+
+This example motivates a number of requirements for the feature. It must be able to
+
+1. specify regions in the context of a global iteration space (for global boundary conditions) using different `with` blocks inside a single stencil
+2. specify cleanly regions at and near the edge of compute domains
+3. read from and write to fields inside halo regions
+
+The specific feature that we are proposing is the addition of a `region()` specifier that complements the `interval()`, but treats the horizontal instead of the vertical.
+Arguments passed to the `region()` specifier are Python expressions resolved at compile time, which result in objects specifying the local index region for the contained computation.
+In order to make this friendlier for applications, they can provide methods or classes that provide these objects specifying the local index region.
+
+Using this concept, the example above is transformed into:
+
+.. code-block:: python
+
+    @gtscript.stencil(backend=backend_type) # Single stencil
+    def ubke(uc: FIELD_TYPE, vc: FIELD_TYPE, cosa: FIELD_TYPE, rsina: FIELD_TYPE, ut: FIELD_TYPE, ub: FIELD_TYPE, dt5: float, dt4: float):
+        with computation(PARALLEL):
+            with interval(...): # applies to the compute domain as it does now
+                ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
+            with region((WestEdge(0, 1), EastEdge(0, 1)), interval(...):
+                # Executes on regions defined by the region objects
+                ub = dt5 * (ut[0, -1, 0] + ut)​
+            with region(SouthEdge(0, 1), NorthEdge(0, 1)), interval(...):
+                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
+
```

Comment:
Let me also try to ask another question that may clarify some of my misunderstandings: is the following code identical to the previous one in terms of output?
```
    @gtscript.stencil(backend=backend_type) # Single stencil
    def ubke(uc: FIELD_TYPE, vc: FIELD_TYPE, cosa: FIELD_TYPE, rsina: FIELD_TYPE, ut: FIELD_TYPE, ub: FIELD_TYPE, dt5: float, dt4: float):
        with computation(PARALLEL):
            with interval(...): # applies to the compute domain as it does now
                ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
            with region(SouthEdge(0, 1), NorthEdge(0, 1)), interval(...):
                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
            with region((WestEdge(0, 1), EastEdge(0, 1)), interval(...):
                # Executes on regions defined by the region objects
                ub = dt5 * (ut[0, -1, 0] + ut)​
```

The same question I would ask if the `intervals` were also non-trivial